### PR TITLE
Speeding up Scala 3 Writing macros

### PR DESCRIPTION
**Pull request by lihaoyi**
Fixes https://github.com/com-lihaoyi/upickle/issues/389 and https://github.com/com-lihaoyi/upickle/issues/388

This PR moves a bunch of logic from runtime to compile time, and optimizes whatever runtime logic remains. It also consolidates the logic with the Scala 2 logic as much as possible, e.g. re-using `writeSnippet` to manage writes and  `CaseObjectContext`/`HugeCaseObjectContext` to manage validation and error reporting during a read

Running some ad-hoc benchmarks on my laptop, `./mill bench.jvm.run`, it gives a significant speedup on reads and writes, and brings it close to the Scala 2.13 numbers (higher is better):

| Benchmark (5000ms)                  | Scala 3 Before | Scala 3 After | Scala 2 |      
|-------------------------------------|----------------|---------------|---------|      
| upickleDefault Read                 |      637       |     1065      |   1403  |      
| upickleDefault Write                |      839       |     1452      |   1549  |      
| upickleDefaultByteArray Read        |      582       |     1172      |   1126  |      
| upickleDefaultByteArray Write       |      847       |     1218      |   1277  |      
| upickleDefaultBinary Read           |      925       |     3853      |   3844  |      
| upickleDefaultBinary Write          |      1252      |     3117      |   3666  |      
| upickleDefaultCached Read           |      620       |     1300      |   1412  |      
| upickleDefaultCached Write          |      829       |     1555      |   1588  |      
| upickleDefaultByteArrayCached Read  |      575       |     1182      |   1095  |      
| upickleDefaultByteArrayCached Write |      838       |     1223      |   1297  |      
| upickleDefaultBinaryCached Read     |      928       |     3825      |   3885  |      
| upickleDefaultBinaryCached Write    |      1266      |     2907      |   3674  |      

Note that the generated code, especially for reads, is still not as optimized as the Scala 2 versions:

1. I couldn't figure out how to generate an anonymous class with typed fields in Scala 3, so I'm putting things in an `Array[Any]`
2. I couldn't figure out how to generate `match` statements, so I generated a `Map[K, V]` and look it up at runtime.

Fixing this issues and moving the reading logic into the macro should also be possible, but can happen in a separate PR

All existing tests pass. Added a regression test for the recursive Scala 3 scenario that hangs on master. Also moved a bunch of `AdvancedTests` into the shared folder now that they work

### CODE REVIEW

File: core/src/upickle/core/Types.scala [hunk 1]
```diff
@@ -277,16 +277,16 @@ trait Types{ types =>
...
-    protected def writeSnippet[R, V](objectAttributeKeyWriteMap: CharSequence => CharSequence,
-                                     ctx: _root_.upickle.core.ObjVisitor[_, R],
-                                     mappedArgsI: String,
-                                     w: Writer[V],
-                                     value: V) = {
+    def writeSnippet[R, V](objectAttributeKeyWriteMap: CharSequence => CharSequence,
+                           ctx: _root_.upickle.core.ObjVisitor[_, R],
+                           mappedArgsI: String,
+                           w: Any,
+                           value: Any) = {
```
Comment (lihaoyi) [inline core/src/upickle/core/Types.scala:284]: Had to make these un-typed because I couldn't figure out how to make the Scala 3 macro quotes/splices play nicely with the path dependent `prefix.Writer` type. A bit gross but shouldn't affect perf or correctness

File: core/src/upickle/core/Types.scala [hunk 2]
```diff
@@ -277,16 +277,16 @@ trait Types{ types =>
         ctx.visitEnd(-1)
       }
     }
-    protected def writeSnippet[R, V](objectAttributeKeyWriteMap: CharSequence => CharSequence,
-                                     ctx: _root_.upickle.core.ObjVisitor[_, R],
-                                     mappedArgsI: String,
-                                     w: Writer[V],
-                                     value: V) = {
+    def writeSnippet[R, V](objectAttributeKeyWriteMap: CharSequence => CharSequence,
```
Comment (lihaoyi) [inline core/src/upickle/core/Types.scala:280]: I couldn't figure out how to subclass the path dependent `prefix.CaseW` type in the macro, so I had to instantiate it in the inline part first and then call into the macro after. That means the call to `writeSnippet` is no longer from the subclass, and it thus can no longer be protected

File: implicits/src-3/upickle/implicits/Readers.scala [hunk 3]
```diff
@@ -1,9 +1,117 @@
 package upickle.implicits
```
Comment (lihaoyi) [inline implicits/src-3/upickle/implicits/Readers.scala:1]: This file is mostly just consolidated from `CaseClassReader.scala`; proper rewrite would be a separate PR

File: implicits/src-3/upickle/implicits/Readers.scala [hunk 4]
```diff
@@ -1,9 +1,110 @@
...
+        macros.checkErrorMissingKeysCount[T]()
+      )
+
+      if macros.isMemberOfSealedHierarchy[T] then annotate(reader, macros.fullClassName[T])
+      else reader
+
+    case m: Mirror.SumOf[T] =>
+      inline compiletime.erasedValue[T] match {
+        case _: scala.reflect.Enum =>
+          val valueOf = macros.enumValueOf[T]
```
Comment (bishabosha) [inline implicits/src-3/upickle/implicits/Readers.scala:92]: this won't work for ADT enums, the solution here is to assert all cases are singleton:
```scala
summonAll[Tuple.Map[m.MirroredElemTypes, [S] =>> S <:< Singleton]] // assert all singleton
```

**Comment by lihaoyi** [comment]
I'll leave this over the weekend, and if there aren't any more comments I'll merge it then. All the existing/new/newly-enabled tests pass, and the benchmarks demonstrate the non-functional improvements, so I can't have broken anything too badly

The issue noticed by @bishabosha is a limitation in existing code, which is why it was not picked up by any tests (since it never worked). Fixing that can come in a follow up

**Review by lefou — APPROVED** [review]
This looks good to me, although I'm myself not experienced in Scala 3 Macros. But tests pass and benchmark looks promising.

After we merge this, I'd like to cut a next milestone release `3.0.0-M2` to get this changed version tested in as many places as possible.

**Comment by lefou** [comment]
There seem to be issues with this change. E.g. in Ammonite, the integration tests don't succeed. 

* https://github.com/com-lihaoyi/Ammonite/pull/1320
